plugins/modules/foreman_hostgroup.py

```diff
@@ -118,31 +211,99 @@ def main():
+            environment=dict(),
+            config_groups=dict(type='list'),
+            puppet_proxy=dict(),
+            puppet_ca_proxy=dict(),
         ),
     )
     entity_dict = module.clean_params()
 
     module.connect()
 
+    # Get short name and parent from provided name
+    name, parent = split_fqn(entity_dict['name'])
+    entity_dict['name'] = name
+
+    if 'parent' in entity_dict:
+        if parent:
+            module.fail_json(msg="Please specify the parent either separately, or as part of the title.")
+        parent = entity_dict['parent']
+    if parent:
+        search_string = 'title="{}"'.format(parent)
+        entity_dict['parent'] = module.find_resource('hostgroups', search=search_string, thin=True, failsafe=module.desired_absent)
+
+        if module.desired_absent and entity_dict['parent'] is None:
+            # Parent hostgroup does not exist so just exit here
+            module.exit_json(changed=False)
+
     if not module.desired_absent:
+        if 'organizations' in entity_dict:
+            entity_dict['organizations'] = module.find_resources_by_name('organizations', entity_dict['organizations'], thin=True)
+
+        if 'locations' in entity_dict:
+            entity_dict['locations'] = module.find_resources_by_title('locations', entity_dict['locations'], thin=True)
+
+        if 'compute_resource' in entity_dict:
+            entity_dict['compute_resource'] = module.find_resource_by_name('compute_resources', name=entity_dict['compute_resource'], failsafe=False, thin=True)
+
+        if 'compute_profile' in entity_dict:
+            entity_dict['compute_profile'] = module.find_resource_by_name('compute_profiles', name=entity_dict['compute_profile'], failsafe=False, thin=True)
+
+        if 'domain' in entity_dict:
+            entity_dict['domain'] = module.find_resource_by_name('domains', name=entity_dict['domain'], failsafe=False, thin=True)
+
+        if 'subnet' in entity_dict:
+            entity_dict['subnet'] = module.find_resource_by_name('subnets', name=entity_dict['subnet'], failsafe=False, thin=True)
+
+        if 'subnet6' in entity_dict:
+            entity_dict['subnet6'] = module.find_resource_by_name('subnets', name=entity_dict['subnet6'], failsafe=False, thin=True)
+
+        if 'realm' in entity_dict:
+            entity_dict['realm'] = module.find_resource_by_name('realms', name=entity_dict['realm'], failsafe=False, thin=True)
+
         if 'architecture' in entity_dict:
             entity_dict['architecture'] = module.find_resource_by_name('architectures', name=entity_dict['architecture'], failsafe=False, thin=True)
 
         if 'operatingsystem' in entity_dict:
-            entity_dict['operatingsystem'] = module.find_resource_by_name('operatingsystems', name=entity_dict['operatingsystem'], failsafe=False, thin=True)
+            entity_dict['operatingsystem'] = module.find_resource_by_title('operatingsystems', title=entity_dict['operatingsystem'], failsafe=False, thin=True)
```

Comment:
Maybe out of scope for this PR, but in other places we use `title ~ STRING` when searching for OSes:
https://github.com/theforeman/foreman-ansible-modules/blob/61fb6c25cfb894a03290192fbd638490b3c8cdae/plugins/modules/foreman_installation_medium.py#L144-L145

https://github.com/theforeman/foreman-ansible-modules/blob/61fb6c25cfb894a03290192fbd638490b3c8cdae/plugins/modules/foreman_architecture.py#L103-L104

While `find_resource_by_title` uses `=`:
https://github.com/theforeman/foreman-ansible-modules/blob/61fb6c25cfb894a03290192fbd638490b3c8cdae/plugins/module_utils/foreman_helper.py#L186-L188

I think the idea behind this was to make the OS search a bit fuzzier, as OS names/titles are weird.

Given you didn't write this code in the first place in this PR, I don't think we need a decision/solution here now, but I wanted to raise awareness for this.

Maybe also @mdellweg can chime in.

Reply:
Oh yes, definitely `~` for comparison, please It seems to come up in more and more places. I know we chatted about that, @evgeni , but how about _some_ support for this in the helper?

Reply:
@evgeni, @mdellweg  I think `~` for comparison to get something can go to not predictable behavior because it can return more than one entity and the code in foreman_helper enforce that we get only one entity when we search an entity (eg searching title ~ CentOS on one of my systems will return 13 entries). I prefer the same behavior as for names, if we don't find any entity with the exact name, it fail (or return None depending on failsafe value).
The title search is mostly used for searching nested resources, operating system is an exception, if we change the helper implementation to use `~` operator, it will break nested resources search.

Reply:
Yeah, the *default* behavior should definitely be `=`. It's just with OperatingSystems we've used fuzzy (`~`) search to find minor versions if the users passes only parts of it (`CentOS 7.3` → `CentOS 7.3.1234`).

Also, `find_resource` will only ever allow the user to get one result and abort if there are muplitple:
https://github.com/theforeman/foreman-ansible-modules/blob/d4e5ce331f1a26fd7319b8f1ff422ec6b4903577/plugins/module_utils/foreman_helper.py#L171-L176

How about we add a flag to `find_resource_by_title` to optionally allow fuzzy (`~`) search. And then add an explicit `find_operatingsystem` that will first try an exact search, and if that fails do a fuzzy one?

Reply:
Ok, so if it's ok for you I would prefer to manage this in another PR that will change how we manage OSes in all places to be consistent and go ahead as is  for now :)

Reply:
Fine for me.

Reply:
Absolutely!

Reply:
I've opened https://github.com/theforeman/foreman-ansible-modules/issues/373 for that.